### PR TITLE
So, I kind of ended up rewriting it...

### PR DESCRIPTION
**Pull request by mnot**

### CODE REVIEW

File: draft-thomson-http-replay.md [hunk 1]
```diff
@@ -34,184 +34,153 @@ informative:
...
 
+2. If the server receives multiple requests in early data, it can determine
+whether to defer HTTP processing on a per-request basis. When doing so, it
+SHOULD defer any requests that have state-changing side effects on the server.
+If this is not known by the server, it MUST defer the request.
 
-# Replay Attacks
+3. When a request received in early data contains the `Early-Data` request
+header field ({{header}}), the server can trigger a retry without the use of
+early data by responding with the 418 (Too Early) status code ({{status)}}), in
```
Comment (martinthomson) [inline draft-thomson-http-replay.md]: 418 is taken. "I'm a teapot".

File: draft-thomson-http-replay.md [hunk 2]
```diff
@@ -34,184 +34,153 @@ informative:
...
 
-For application data that is sent after the TLS handshake completes, TLS
-provides protection against replay.  Duplicated data is detected and treated as
-a fatal error condition (in DTLS {{?DTLS=RFC6347}}, duplicated data is instead
-discarded).  Similarly, transport layer loss recovery, such as TCP
-retransmission, does not generate replays because the transport discards
-duplicated data.
+Note that a server cannot choose to selectively reject early data. TLS only
+permits a server to accept all early data, or none of it. Once a server has
+decided to accept early data, it MUST process all requests in early data.
```
Comment (martinthomson) [inline draft-thomson-http-replay.md]: I think that we should add: "... Even if the server rejects the request by sending a 418/422 response."

File: draft-thomson-http-replay.md [hunk 3]
```diff
@@ -34,184 +34,153 @@ informative:
...
 
-TLS early data creates the potential for a replay attack that cannot be detected
-by TLS or the transport protocol.  TLS {{?TLS13}} describes several mitigation
-strategies that reduce the ability of an attacker to successfully replay early
-data.  Servers are strongly encouraged to implement these techniques, but to
-also recognize that they are imperfect.
+Also, a server can limit the amount of early data with the `max_early_data_size`
```
Comment (martinthomson) [inline draft-thomson-http-replay.md]: Remove "also".

File: draft-thomson-http-replay.md [hunk 4]
```diff
@@ -220,54 +189,45 @@ Early-Data = "0" / "1"
...
-until the TLS handshake completes and forward the request again.
 
 
-## The 422 (Too Early) Status Code
+## The 418 (Too Early) Status Code {#status}
 
-A 422 (Too Early) status code indicates that the server that is unwilling to
-risk processing requests that might be replayed.
+A 418 (Too Early) status code indicates that the server is unwilling to risk
+processing a request that might be (or have been) replayed.
```
Comment (martinthomson) [inline draft-thomson-http-replay.md]: has been

**Review by martinthomson — CHANGES_REQUESTED** [review]
I'd merge this if not for two things:
1. 418 is taken (we told IANA about this year's ago, but the registry is still not fixed, I am going to blame you for that)
2. The text on retries you promised is missing. You need to say explicitly that this talks about replays and not the retries that might be initiated by clients, automatic or otherwise.

**Comment by mnot** [comment]
1. No, 418 is not taken; HTCP is not HTTP. Different name space. 
2. OK, I'll add a mention.

**Comment by mnot** [comment]
BTW, 422 *is* taken.

**Comment by mnot** [comment]
One thing to note -- I explicitly talk about not putting safe requests into early data. Might want to raise an issue on that so we can discuss it (I'm happy to be talked around).

**Comment by wtarreau** [comment]
Mark, over the last 2 decades, 418 has been deployed in real life : 
- http://www.google.com/teapot
- http://teapot.joereddington.com/
- http://error418.net/
- https://meta.stackexchange.com/questions/185426/stack-overflow-returning-http-error-code-418-im-a-teapot
- it's even listed on the wikipedia's list of official status codes (as you know, developers use wikipedia as an official protocol source).
And apparently some search engines even make a special case of it : https://sitesdoneright.com/blog/2013/03/what-is-418-im-a-teapot-status-code-error

So OK it was a successful prank, Larry did a good job at it, but whether we like it or not, it's being "used". 

I couldn't find any trace of 427 however, which is marked as unassigned on the IANA, so let's pick this one ?

**Comment by mnot** [comment]
If we let a couple of sites deploying a joke status code prevent us from using a scarce resource (we only have so many left), we might as well pack it in. Using 418 in this manner won't cause any interoperability problems with a *real* protocol.

**Comment by mnot** [comment]
BTW, the *official* advice that some *official* people came up with in RFC7231 is that -- at this point - -we should be using something like `4NN`. Should I just change it to that for now, and let the WG fight it out?

**Pull request merged**